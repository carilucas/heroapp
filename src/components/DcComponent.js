import React from 'react'
import { HeroList } from './heros/HeroList'

export const DcComponent = () => {
   return (
      <div>
         <h1>DcComponent</h1>
         <hr/>
         <HeroList publisher='DC Comics' />
      </div>
   )
}
