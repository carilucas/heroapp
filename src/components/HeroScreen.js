import React, { useContext } from 'react'
import { Redirect, useParams } from 'react-router';
import { getHeroById } from '../helpers/getHeroById';
import {AuthCotext} from '../auth/authContext'




export const HeroScreen = ({history}) => {
   const {heroId} = useParams();
   const hero = getHeroById(heroId);

   const {heroImages} = useContext(AuthCotext)
   
   if (!hero) {
      
      return <Redirect to="/" />
   }

   const {id,superhero,publisher,alter_ego,first_appearance} = hero;
   

   const handleReturn = ()=>{
      if (history.length <= 2) {
         history.push('/')
      }else{
         history.goBack();
      }
   }
   return (
      <div>
         <h1>{superhero} </h1>
         <hr/>
         <div className="row">
            <div className="col-sm-4 animate__animated animate__fadeIn">
               <img 
                  // src={`../assets/heroes/${id}.jpg`} 
                  src={ heroImages( `./${id}.jpg` ).default }
                  alt={superhero} 
                  className="img-fluid" 
               />
            </div>
            <div className="col-sm-8 animate__animated animate__fadeIn">
            <ul className="list-group">
               <li className="list-group-item">Publisher: {publisher}</li>
               <li className="list-group-item">Alter Ego: {alter_ego}</li>
               <li className="list-group-item">First appearance: {first_appearance}</li>
               <button
                  className="btn btn-primary mt-3"
                  onClick={ handleReturn }
               >
                  Go back
               </button>
            </ul>
            </div>
         </div>
      </div>
   )
}
