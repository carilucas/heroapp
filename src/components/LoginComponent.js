import React, { useContext } from 'react';
import { AuthCotext } from '../auth/authContext';
import { LOGIN } from '../auth/types';



export const LoginComponent = ({history}) => {

   const { dispatchUser } = useContext(AuthCotext);
   
   const lastPath = localStorage.getItem('lastPath') || '/';

   
   const handleLogin = ()=>{
      dispatchUser({
         type: LOGIN,
         payload:{
            name:'Carlos',
            logged: true
         }
      })
      history.push(lastPath)
   }



   return (
      <div className="container mt-5">
         <h1>LoginComponent</h1>
         <hr/>
         <button 
            className="btn btn-primary"
            onClick={ handleLogin }
         >
            Login
         </button>
      </div>
   )
}
