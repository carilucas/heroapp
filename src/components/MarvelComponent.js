import React from 'react'
import { HeroList } from './heros/HeroList'

export const MarvelComponent = () => {
   return (
      <div>
         <h1>MarvelComponent</h1>
         <hr/>
         <HeroList publisher='Marvel Comics' />
      </div>
   )
}
