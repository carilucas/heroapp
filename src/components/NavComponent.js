import React, { useContext } from 'react'
import { Link, NavLink, useHistory } from 'react-router-dom'
import { AuthCotext } from '../auth/authContext'
import { LOGOUT } from '../auth/types'

export const Navbar = () => {
      const history = useHistory();
      const { user:{name}, dispatchUser } = useContext(AuthCotext);

      
      
      const handleLogout = ()=>{
            history.replace('/login');
            dispatchUser({
                  type: LOGOUT
            });
            
      }

    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">         
            <div className="container">
               <Link 
                  className="navbar-brand" 
                  to="/"
               >
                  Hero App
               </Link>

               <div className="navbar-collapse">
                  <div className="navbar-nav">

                     <NavLink 
                           activeClassName="active"
                           className="nav-item nav-link" 
                           exact
                           to="/marvel"
                     >
                           Marvel
                     </NavLink>

                     <NavLink 
                           activeClassName="active"
                           className="nav-item nav-link" 
                           exact
                           to="/dc"
                     >
                           DC
                     </NavLink>
                     
                     <NavLink 
                           activeClassName="active"
                           className="nav-item nav-link" 
                           exact
                           to="/search"
                     >
                           Search 
                     </NavLink>
                  </div>
               </div>

               <div className="navbar-collapse collapse w-100 order-3 dual-collapse2">
                  <ul className="navbar-nav ms-auto">
                        <li>
                              <span className="nav-item nav-link" style={{color:'cyan'}}>{name}</span>
                        </li>
                        <li>
                              <button 
                                    className="nav-item nav-link btn btn-outline-info" 
                                    onClick={ handleLogout }
                              >
                                    Logout
                              </button>
                        </li>
                  </ul>
               </div>
            </div>
        </nav>
    )
}