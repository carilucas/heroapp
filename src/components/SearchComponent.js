import React, { useMemo } from 'react';
import { useForm } from '../hooks/useForm';
import queryString  from 'query-string';
import { getHeroByName } from '../helpers/getHeroByName';
import { HeroListItem } from './heros/HeroListItem';
import { useLocation } from 'react-router';

export const SearchComponent = ({history}) => {
   
   const location = useLocation();
   
   const {hero=''} = queryString.parse( location.search );

   const heros = useMemo(() => getHeroByName(hero), [hero]);

   const [value, handleInputChange ] = useForm({searchText:hero});
   const {searchText} = value; 
   
   const handleSubmit = (e)=>{
      e.preventDefault();

      history.push(`?hero=${searchText}`)
      
   }


   return (
      <div>
         <h1>Search</h1>
         <hr/>
         <div className="row">
            <div className="col-sm-3">
               <form 
                  onSubmit = { handleSubmit }
                  
               >
                  <input 
                     type="text" 
                     className="form-control"
                     name="searchText"
                     onChange = { handleInputChange }
                     autoComplete="off"
                     value={ searchText }
                  />
                  <button
                     className="btn btn-primary mt-3"
                  >
                     Search
                  </button>
               </form>
            </div>
            <div className="col-sm-9">
               <div className="row animate__animated animate__fadeIn">
                  {
                     hero === '' 
                        && <div className="alert alert-info">Enter a hero name</div>
                  }
                  {
                     hero !== '' && heros.length < 1
                        && <div className="alert alert-danger">There is no matches for {hero}</div>
                  }
                  {
                     heros.map( hero => (
                        <HeroListItem {...hero} key={hero.id}/>
                     ))
                  }
               </div>
            </div>
         </div>
      </div>
   )
}
