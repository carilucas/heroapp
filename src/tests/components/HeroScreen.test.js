import { HeroScreen } from "../../components/HeroScreen"
import { mount } from 'enzyme'
import { MemoryRouter, Route } from "react-router";

describe('Pruebas en <HeroScreen>', () => {

   const history = {
      length : 10,
      goBack: jest.fn(),
      push: jest.fn(),
   }


   test('Mostrar el snapshot del redirect si no hay argumentos en el url', () => {
      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero']}>
            <HeroScreen history={ history }/>
         </MemoryRouter>
         
      );
      expect(wrapper).toMatchSnapshot();
      expect( wrapper.find('Redirect').exists()).toBe(true)
   });

   test('Debe mostrar un heroe si el parÃ¡metro existe y se encuentra', () => {
      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Route path='/hero/:heroId' component={HeroScreen} />
         </MemoryRouter>
      );

      expect( wrapper.find('.row').exists()).toBe(true);
   });
   
   test('Debe regresar a la pantalla anterior con push', () => {

      const history = {
         length : 1,
         goBack: jest.fn(),
         push: jest.fn(),
      }

      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Route 
               path='/hero/:heroId' 
               component={(props)=><HeroScreen history={ history }/>} 
            />   
         </MemoryRouter>
      );

      wrapper.find('button').prop('onClick')();

      expect(history.push).toHaveBeenCalled();
      expect(history.push).toHaveBeenCalledWith('/');
      expect(history.goBack).not.toHaveBeenCalled();

   });

   test('Debe llamar a la funcion goBack', () => {
      
      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Route 
               path='/hero/:heroId'
               component={ (props)=> <HeroScreen history={history} />}
            />
         </MemoryRouter>

      );

      wrapper.find('button').prop('onClick')();
      
      expect(history.goBack).toHaveBeenCalled();
      expect(history.push).toHaveBeenCalledTimes(0);


   });

   test('Debe redireccionar si el parametro id no se encuentra', () => {
      
      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero/marvel-spiderwww']}>
            <Route 
               path='/hero/:heroId'
               component={ (props)=> <HeroScreen history={history} />}
            />
         </MemoryRouter>
      );

      console.log(wrapper.html());

      expect( wrapper.html() ).toBe('');

   })
   
   
   
   
})
