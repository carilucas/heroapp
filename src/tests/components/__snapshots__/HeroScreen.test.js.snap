// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen> Mostrar el snapshot del redirect si no hay argumentos en el url 1`] = `""`;
